### DISCUSSION TITLE
tutorials/best_practices/data_preferences

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#50] [User-contributed notes]
# tutorials/best_practices/data_preferences

Ever wondered whether one should approach problem X with data structure Y or Z? This article covers a variety of topics related to these dilemmas. Array vs. Dictionary vs. Object: Godot stores all ...

https://docs.godotengine.org/en/stable/tutorials/best_practices/data_preferences.html

**Comment by sebbasim** [comment]
Enumerations: int vs. string
If I understand correctly, if you use @export_enum, you lose type checking. This is an important consideration, and should be included in the paragraph.

**Comment by Clark-E** [comment]
The details about how Dictionaries are implemented seem to be incorrect or outdated. Unless I'm mistaken, Dictionaries sizes *aren't* powers of 2, but instead use the sizes defined by ```hash_table_size_primes``` in ```core\templates\hashfuncs.h```. Also, Dictionaries don't use ```OrderedHashMap``` (which isn't in the codebase,) but instead use ```HashMap<Variant, Variant, VariantHasher, StringLikeVariantComparator>```.

**Comment by Phanterm** [comment]
It is worth mentioning that one can access an existing function on named enums by calling ``MyEnum.keys()[]``, passing into the brackets the value of any property using this enum. One need not write their own Dictionary for this as is stated on this page. For example:

```
enum RiceType {
    Jasmine,
    Basmati,
    Calrose
}

var favorite_rice : RiceType = RiceType.Jasmine
```

```
print(favorite_rice)
```
Because ``favorite_rice = RiceType.Jasmine``, this line will print ``0``, as Jasmine occupies index ``0`` of our ``RiceType`` enum.

```
print(RiceType.keys()[favorite_rice])
```
This line will print ``Jasmine``. Because we are accessing the Dictionary stored on the ``RiceType`` enum, and by passing in ``favorite_rice`` (which, remember, is equal to ``0`` because Jasmine is at index `0`), we return the key at that position.

**Reply by elliotfontaine** [reply]
Sadly, this only work if you don't override the default indexing when setting up the enum. For example, the following would cause an out of bound error when calling `print(RiceType.keys()[favorite_rice])` :
```gdscript
enum RiceType {
    Jasmine = 10,
    Basmati = 1,
    Calrose = 2
}
```

**Comment by xmamalou** [comment]
Note the following on Arrays vs Hashmaps:
- One generally needs an insanely high amount of data for hashmaps to prove better, in terms of performance, than arrays (this is true for many other sophisticated data structures as well, like linked lists). Most array accesses and traversals are not really that slow, as thanks to their contiguous nature they are most likely already cached by the CPU. Looping over an array is also fast after a certain point, as the branch predictor of the CPU will most likely consider taking the loop after a while.
- Moving data in an array is actually *cheap*. Since we work with virtual memory, and not physical memory, the only thing moving data to a new location needs is to actually change the virtual address of that data. (Sidenote: the CPU cache itself is not concerned on where things lie in physical memory. It only cares about where they lie in virtual memory. Hence, it's very likely that things that appear contiguous in virtual memory are actually rather fragmented in physical memory). Not only that, but adding elements to an array will not require a lot of allocations if the array itself preallocates some capacity for future objects. On the other hand, if a hash of a hashmap proves to be outside the already allocated memory (which is not unlikely, as hashes are supposed to look random), this implies a lot more necessary allocations on average (unless the hashmap preallocates an absurd amount of memory or does some other magic). And allocating memory is *super slow*.
- Calculating hashes to find an element is actually *slow* compared to traversing an array to find it, especially if that place is already stored in the CPU cache, which is very likely if the array is accessed often, despite the fact that the former operation is `O(1)` whilst the latter is `O(n)`. Generally, one should refrain from making definitive performance judgements based on the big-O notation. That notation concerns the performance of an algorithm *when its input tends to infinity* (hence the comment on the first bullet about needing an insanely high amount of data). 

Arrays prove themselves great for a lot of tasks, despite common intuition shutting them off as slow and impractical. Unless a profiler really tells you that you do need a hashmap for something, please prefer using arrays for storing data instead! (one should also consider breaking the elements an object into their own separate arrays when they want an array of that object. So, if you had an object with 2 members, instead of having an array with instances of that object, you can have two arrays with instances of one of its members each)

**Reply by Calinou** [reply]
Did you profile this on Godot's data structures? Remember that Godot does not use STL containers, but its own implementations as described in [Core types](https://docs.godotengine.org/en/latest/engine_details/architecture/core_types.html).
